### ISSUE TITLE
No mention on how to provide input file

### ISSUE THREAD

**Issue by cdgriffith** [wswartzendruber/pq2hlg#1]
In the Readme and some quick testing there doesn't seem to be a way to provide an input file? 

I also tried piping to STDIN like other programs do `ffmpeg -i "input.mkv" -c:v copy -vbsf hevc_mp4toannexb -f hevc - | cargo run . out.cube` and did not work

**Comment by wswartzendruber** [comment]
It doesn't take input files. It takes parameters. I will see about updating the README over the weekend, but you will also need the pqstat tool and the output it generates.

**Comment by wswartzendruber** [comment]
I did not get the README created over the weekend. In fact, I found out that consumer HDR video is limited RGB, not full RGB, so I need to go back and make the table generation algorithm configurable in this manner.

**Comment by wswartzendruber** [comment]
All right, the project is tentatively complete. Read over the README.md file and let me know if you still have any questions.

**Comment by cdgriffith** [comment]
So for a workflow of how to do this would be first run `pqstat`: 

```
ffmpeg -hide_banner -v quiet -i "ducks..mp4" -map 0:0 -c copy -f rawvideo -pix_fmt yuv420p10le - | pqstat.exe  --height 1920 --width 1080 -
```

Which will provide: 
```
MaxCLL....: 9912.010258970728
MaxChannel: 1
```

Put that through `pq2hlg`
```
pq2hlg.exe -m 1 -r 9912.010258970728 out.cube
```

And then use that with `lut3d` filter via ffmpeg, specifying it to be HLG via `arib-std-b67`:

```
ffmpeg -i "ducks.mp4" -vf lut3d="out.cube" -c:v libx265 -pix_fmt yuv420p10le -x265-params "colorprim=bt2020:transfer=arib-std-b67:colormatrix=bt2020nc" test.mkv
```

The problem I run into is that it is now a lot darker.
![lut](https://user-images.githubusercontent.com/3275435/107545178-9e495200-6b90-11eb-8a63-ee109ea9580b.png)

**Comment by wswartzendruber** [comment]
You have set the reference white level (-r) to what MaxCLL is. This will cause MaxCLL to be driven down (in brightness) to HLG's standard reference white level, along with everything in between. As reference white is essentially how bright a piece of paper would be, and MaxCLL is the brightest pixel encountered, this is almost certainly **not** what you want.

Unfortunately, I do not have an automated way of determining where reference white is in a given video source. SMPTE says it should be at 100 nits. The ITU says it should be at 203 nits. When it comes to UltraHD Blu-rays, they are anywhere in between. This is why pq2hlg lets you configure it.

Now, the main reason reference white is so important is that it aids greatly in how your HLG output looks in SDR. So I would start with the default value of 203 nits and then preview it in your favorite SDR player. If the picture is too dim, lower the -r value. If it's too bright, increase the -r value.

**Comment by wswartzendruber** [comment]
By the way, a trick I use is to find a scene that doesn't have any specular brightness in it, but that does have something like a piece of paper or a cotton shirt. I then put this frame (by itself) through pqstat and take the MaxCLL value of that frame to be reference white. Or, you can find a frame with specular detail and crop that out leaving only the paper or white shirt.

**Comment by wswartzendruber** [comment]
Also, I missed that you're feeding YUV420P10LE into pqstate. That tool expects RGB48LE.

ffmpeg ... -vf format=rgb48le ...

**Comment by cdgriffith** [comment]
Ah good catch, though it is still giving the really high value of 9912:
```
ffmpeg -hide_banner -v quiet -i "ducks.mp4" -map 0:0 -c copy -f rawvideo -pix_fmt rgb48le - | pqstat.exe  --height 1920 --width 1080 -
MaxCLL....: 9912.010258970728
MaxChannel: 1
```

via FFprobe as well as MediaInfo it is giving: 
```
Maximum Content Light Level              : 1000 cd/m2
Maximum Frame-Average Light Level        : 200 cd/m2
```

I tried with both of those values, and 200 gave me what was looking identical to the source. Video is from a Samsung phone, so it may just be sticking to that 203 nits spec as well, hence the same look.

**Comment by wswartzendruber** [comment]
My last comment was a bit late, but you need to use `-vf format=rgb48le` to force a signal change.

Maximum Content Light Level = MaxCLL
Maximum Frame-Average Light Level = MaxFALL

Neither are reference white.

**Comment by cdgriffith** [comment]
> You have set the reference white level (-r) to what MaxCLL is...

> Unfortunately, I do not have an automated way of determining where reference white is in a given video source.

Guess I am just a bit confused. To be clear, there is no need to run pqstat tool, as that does NOT give us that `-r` number? For now is a best guess?

(Updated output just to show number is now a bit lower, sorry I totally missed it was a filter!) 
```
ffmpeg -hide_banner -v quiet -i "ducks.mp4" -map 0:0 -vf format=rgb48le -f rawvideo - | pqstat.exe  --height 1920 --width 1080 -
MaxCLL....: 9504.738319386346
MaxChannel: 1
```

**Comment by wswartzendruber** [comment]
You still need `pqstat` to give you the MaxChannel value. This is so that pq2hlg tone maps the PQ video to the exact extent needed, no more, no less.

In your case, MaxChannel is 1. That seems......high. But your FFMPEG command line also looks correct. I'll come back to this after work.

**Comment by wswartzendruber** [comment]
I just don't get why it's telling you that. Can you share `ducks.mp4`?

**Comment by cdgriffith** [comment]
https://github.com/cdgriffith/FastFlix/tree/master/tests/media   full name of "Beverly Hills Duck Pond - HDR10plus - Jessica Payne.mp4"

**Comment by wswartzendruber** [comment]
```
MaxCLL....: 9504.738319386346
MaxChannel: 1
```

The difference could be the decoder I used, but that's still disturbingly high. The metadata you read from FFprobe is too rounded and whole to be real. I think the camera app hardcodes those values in.

I've parsed out the HDR10+ metadata, but I'm not familiar with what it all means.

I have unit tests for `pqstat` that ensure that various RGB48LE values return the appropriate readings.

The thing is, I don't think there's a camera out there that can reliably record 9,000+ nits. Could I convince you to submit this sample to the people on Doom9 and ask them evaluate it?
